File: universal-transport/swift/Sources/UniversalTransportSharedMemory/BinaryProtocol.swift
//
//  BinaryProtocol.swift
//  High-Performance Binary Protocol for Swift
//
//  TCP-like fixed binary protocol matching Rust implementation exactly
//

import Foundation

// MARK: - Protocol Constants

/// Protocol magic number (matches Rust) - "UTPB"
public let BINARY_PROTOCOL_MAGIC: UInt32 = 0x55545042

/// Protocol version
public let BINARY_PROTOCOL_VERSION: UInt8 = 1

/// Message header size (32 bytes, cache-line aligned)
public let BINARY_HEADER_SIZE = 32

/// Maximum payload size (64MB)
public let BINARY_MAX_PAYLOAD_SIZE: UInt32 = 64 * 1024 * 1024

// MARK: - Message Types

/// Message types (matches Rust enum exactly)
public enum BinaryMessageType: UInt8, CaseIterable {
    case data = 0x01
    case heartbeat = 0x02
    case acknowledgment = 0x03
    case error = 0x04
    case benchmark = 0x05
}

// MARK: - Binary Message Header

/// Binary message header (32 bytes, exact match with Rust)
/// Layout matches Rust BinaryHeader exactly:
/// 0-3:   Magic number (4 bytes)
/// 4:     Version (1 byte) 
/// 5:     Message type (1 byte)
/// 6-7:   Flags (2 bytes)
/// 8-11:  Payload length (4 bytes)
/// 12-19: Sequence number (8 bytes)
/// 20-27: Timestamp (8 bytes)
/// 28-31: CRC32 checksum (4 bytes)
public struct BinaryMessageHeader {
    public var magic: UInt32
    public var version: UInt8
    public var messageType: UInt8
    public var flags: UInt16
    public var payloadLength: UInt32
    public var sequence: UInt64
    public var timestamp: UInt64
    public var checksum: UInt32
    
    public init(messageType: BinaryMessageType, payload: Data) {
        self.magic = BINARY_PROTOCOL_MAGIC
        self.version = BINARY_PROTOCOL_VERSION
        self.messageType = messageType.rawValue
        self.flags = 0
        self.payloadLength = UInt32(payload.count)
        self.sequence = 0 // Set by sender
        self.timestamp = UInt64(Date().timeIntervalSince1970 * 1_000_000) // microseconds
        self.checksum = Self.calculateCRC32(for: payload)
    }
    
    /// Set sequence number
    public mutating func setSequence(_ seq: UInt64) {
        self.sequence = seq
    }
    
    /// Validate header
    public func validate() throws {
        guard magic == BINARY_PROTOCOL_MAGIC else {
            throw BinaryProtocolError.invalidMagic(magic)
        }
        
        guard version == BINARY_PROTOCOL_VERSION else {
            throw BinaryProtocolError.unsupportedVersion(version)
        }
        
        guard payloadLength <= BINARY_MAX_PAYLOAD_SIZE else {
            throw BinaryProtocolError.payloadTooLarge(payloadLength)
        }
    }
    
    /// Verify payload checksum
    public func verifyChecksum(for payload: Data) -> Bool {
        return checksum == Self.calculateCRC32(for: payload)
    }
    
    /// Calculate CRC32 checksum (matches Rust implementation)
    private static func calculateCRC32(for data: Data) -> UInt32 {
        // Simple CRC32 implementation for compatibility
        var crc: UInt32 = 0xFFFFFFFF
        
        for byte in data {
            crc ^= UInt32(byte)
            for _ in 0..<8 {
                if (crc & 1) != 0 {
                    crc = (crc >> 1) ^ 0xEDB88320
                } else {
                    crc = crc >> 1
                }
            }
        }
        
        return ~crc
    }
    
    /// Serialize header to bytes (little-endian, matches Rust)
    public func toBytes() -> Data {
        var data = Data(capacity: BINARY_HEADER_SIZE)
        
        // Magic (4 bytes, little-endian)
        withUnsafeBytes(of: magic.littleEndian) { data.append(contentsOf: $0) }
        
        // Version (1 byte)
        data.append(version)
        
        // Message type (1 byte) 
        data.append(messageType)
        
        // Flags (2 bytes, little-endian)
        withUnsafeBytes(of: flags.littleEndian) { data.append(contentsOf: $0) }
        
        // Payload length (4 bytes, little-endian)
        withUnsafeBytes(of: payloadLength.littleEndian) { data.append(contentsOf: $0) }
        
        // Sequence (8 bytes, little-endian)
        withUnsafeBytes(of: sequence.littleEndian) { data.append(contentsOf: $0) }
        
        // Timestamp (8 bytes, little-endian)
        withUnsafeBytes(of: timestamp.littleEndian) { data.append(contentsOf: $0) }
        
        // Checksum (4 bytes, little-endian)
        withUnsafeBytes(of: checksum.littleEndian) { data.append(contentsOf: $0) }
        
        return data
    }
    
    /// Deserialize header from bytes
    public static func fromBytes(_ data: Data) throws -> BinaryMessageHeader {
        guard data.count >= BINARY_HEADER_SIZE else {
            throw BinaryProtocolError.insufficientData(data.count)
        }
        
        return try data.withUnsafeBytes { bytes in
            var offset = 0
            
            // Magic (4 bytes, little-endian)
            let magic = bytes.loadUnaligned(fromByteOffset: offset, as: UInt32.self).littleEndian
            offset += 4
            
            // Version (1 byte)
            let version = bytes[offset]
            offset += 1
            
            // Message type (1 byte)
            let messageType = bytes[offset]
            offset += 1
            
            // Flags (2 bytes, little-endian)
            let flags = bytes.loadUnaligned(fromByteOffset: offset, as: UInt16.self).littleEndian
            offset += 2
            
            // Payload length (4 bytes, little-endian)
            let payloadLength = bytes.loadUnaligned(fromByteOffset: offset, as: UInt32.self).littleEndian
            offset += 4
            
            // Sequence (8 bytes, little-endian)
            let sequence = bytes.loadUnaligned(fromByteOffset: offset, as: UInt64.self).littleEndian
            offset += 8
            
            // Timestamp (8 bytes, little-endian)
            let timestamp = bytes.loadUnaligned(fromByteOffset: offset, as: UInt64.self).littleEndian
            offset += 8
            
            // Checksum (4 bytes, little-endian)
            let checksum = bytes.loadUnaligned(fromByteOffset: offset, as: UInt32.self).littleEndian
            
            let header = BinaryMessageHeader(
                magic: magic,
                version: version,
                messageType: messageType,
                flags: flags,
                payloadLength: payloadLength,
                sequence: sequence,
                timestamp: timestamp,
                checksum: checksum
            )
            
            try header.validate()
            return header
        }
    }
    
    private init(magic: UInt32, version: UInt8, messageType: UInt8, flags: UInt16,
                 payloadLength: UInt32, sequence: UInt64, timestamp: UInt64, checksum: UInt32) {
        self.magic = magic
        self.version = version
        self.messageType = messageType
        self.flags = flags
        self.payloadLength = payloadLength
        self.sequence = sequence
        self.timestamp = timestamp
        self.checksum = checksum
    }
}

// MARK: - Complete Binary Message

/// Complete binary message (header + payload)
public struct BinaryMessage {
    public var header: BinaryMessageHeader
    public let payload: Data
    
    public init(messageType: BinaryMessageType, payload: Data) throws {
        guard payload.count <= BINARY_MAX_PAYLOAD_SIZE else {
            throw BinaryProtocolError.payloadTooLarge(UInt32(payload.count))
        }
        
        self.header = BinaryMessageHeader(messageType: messageType, payload: payload)
        self.payload = payload
    }
    
    /// Create a benchmark message
    public static func benchmark(id: UInt64, data: Data) throws -> BinaryMessage {
        var message = try BinaryMessage(messageType: .benchmark, payload: data)
        message.header.setSequence(id)
        return message
    }
    
    /// Get total message size
    public var totalSize: Int {
        return BINARY_HEADER_SIZE + payload.count
    }
    
    /// Validate complete message
    public func validate() throws {
        try header.validate()
        
        guard header.verifyChecksum(for: payload) else {
            throw BinaryProtocolError.checksumMismatch
        }
    }
    
    /// Serialize entire message to bytes
    public func toBytes() -> Data {
        var data = Data(capacity: totalSize)
        data.append(header.toBytes())
        data.append(payload)
        return data
    }
    
    /// Deserialize message from bytes
    public static func fromBytes(_ data: Data) throws -> BinaryMessage {
        guard data.count >= BINARY_HEADER_SIZE else {
            throw BinaryProtocolError.insufficientData(data.count)
        }
        
        // Parse header
        let headerData = data.prefix(BINARY_HEADER_SIZE)
        let header = try BinaryMessageHeader.fromBytes(headerData)
        
        // Check payload size
        let expectedTotal = BINARY_HEADER_SIZE + Int(header.payloadLength)
        guard data.count >= expectedTotal else {
            throw BinaryProtocolError.insufficientData(data.count)
        }
        
        // Extract payload
        let payload = data.subdata(in: BINARY_HEADER_SIZE..<expectedTotal)
        
        let message = BinaryMessage(header: header, payload: payload)
        try message.validate()
        
        return message
    }
    
    private init(header: BinaryMessageHeader, payload: Data) {
        self.header = header
        self.payload = payload
    }
}

// MARK: - Benchmark Message

/// Benchmark-specific message (matches Rust implementation exactly)
public struct BinaryBenchmarkMessage {
    public let id: UInt64
    public let timestamp: UInt64
    public let data: Data
    public let metadata: String
    
    public init(id: UInt64, dataSize: Int) {
        self.id = id
        self.timestamp = UInt64(Date().timeIntervalSince1970 * 1_000_000) // microseconds
        self.data = Data(repeating: 0x42, count: dataSize)
        self.metadata = "benchmark_msg_\(id)"
    }
    
    /// Serialize to binary format (exact match with Rust layout)
    /// Layout:
    /// 0-7:    ID (8 bytes, little-endian)
    /// 8-15:   Timestamp (8 bytes, little-endian)
    /// 16-19:  Data length (4 bytes, little-endian)
    /// 20-23:  Metadata length (4 bytes, little-endian)
    /// 24+:    Data
    /// N+:     Metadata (UTF-8)
    public func toBinary() -> Data {
        let metadataBytes = metadata.data(using: .utf8) ?? Data()
        let totalSize = 24 + data.count + metadataBytes.count
        
        var result = Data(capacity: totalSize)
        
        // ID (8 bytes, little-endian)
        withUnsafeBytes(of: id.littleEndian) { result.append(contentsOf: $0) }
        
        // Timestamp (8 bytes, little-endian)  
        withUnsafeBytes(of: timestamp.littleEndian) { result.append(contentsOf: $0) }
        
        // Data length (4 bytes, little-endian)
        withUnsafeBytes(of: UInt32(data.count).littleEndian) { result.append(contentsOf: $0) }
        
        // Metadata length (4 bytes, little-endian)
        withUnsafeBytes(of: UInt32(metadataBytes.count).littleEndian) { result.append(contentsOf: $0) }
        
        // Data
        result.append(data)
        
        // Metadata
        result.append(metadataBytes)
        
        return result
    }
    
    /// Deserialize from binary format
    public static func fromBinary(_ data: Data) throws -> BinaryBenchmarkMessage {
        guard data.count >= 24 else {
            throw BinaryProtocolError.insufficientData(data.count)
        }
        
        return try data.withUnsafeBytes { bytes in
            var offset = 0
            
            // ID (8 bytes, little-endian)
            let id = bytes.loadUnaligned(fromByteOffset: offset, as: UInt64.self).littleEndian
            offset += 8
            
            // Timestamp (8 bytes, little-endian)
            let timestamp = bytes.loadUnaligned(fromByteOffset: offset, as: UInt64.self).littleEndian
            offset += 8
            
            // Data length (4 bytes, little-endian)
            let dataLen = Int(bytes.loadUnaligned(fromByteOffset: offset, as: UInt32.self).littleEndian)
            offset += 4
            
            // Metadata length (4 bytes, little-endian)
            let metadataLen = Int(bytes.loadUnaligned(fromByteOffset: offset, as: UInt32.self).littleEndian)
            offset += 4
            
            // Check remaining data
            guard data.count >= offset + dataLen + metadataLen else {
                throw BinaryProtocolError.insufficientData(data.count)
            }
            
            // Extract data
            let messageData = data.subdata(in: offset..<(offset + dataLen))
            offset += dataLen
            
            // Extract metadata
            let metadataData = data.subdata(in: offset..<(offset + metadataLen))
            guard let metadata = String(data: metadataData, encoding: .utf8) else {
                throw BinaryProtocolError.invalidUtf8
            }
            
            return BinaryBenchmarkMessage(id: id, timestamp: timestamp, data: messageData, metadata: metadata)
        }
    }
    
    private init(id: UInt64, timestamp: UInt64, data: Data, metadata: String) {
        self.id = id
        self.timestamp = timestamp
        self.data = data
        self.metadata = metadata
    }
    
    /// Convert to binary message
    public func toBinaryMessage() throws -> BinaryMessage {
        let payload = toBinary()
        return try BinaryMessage.benchmark(id: id, data: payload)
    }
    
    /// Create from binary message
    public static func fromBinaryMessage(_ message: BinaryMessage) throws -> BinaryBenchmarkMessage {
        return try fromBinary(message.payload)
    }
}

// MARK: - Protocol Errors

public enum BinaryProtocolError: Error, LocalizedError {
    case invalidMagic(UInt32)
    case unsupportedVersion(UInt8)
    case payloadTooLarge(UInt32)
    case insufficientData(Int)
    case checksumMismatch
    case invalidUtf8
    
    public var errorDescription: String? {
        switch self {
        case .invalidMagic(let magic):
            return "Invalid magic number: 0x\(String(magic, radix: 16))"
        case .unsupportedVersion(let version):
            return "Unsupported protocol version: \(version)"
        case .payloadTooLarge(let size):
            return "Payload too large: \(size) bytes"
        case .insufficientData(let available):
            return "Insufficient data: \(available) bytes available"
        case .checksumMismatch:
            return "Checksum mismatch"
        case .invalidUtf8:
            return "Invalid UTF-8 encoding"
        }
    }
}

File: universal-transport/rust/core/src/binary_protocol.rs
//! High-Performance Binary Protocol for Universal Transport
//! 
//! TCP-like fixed-length binary protocol for maximum performance
//! No JSON overhead - direct memory serialization

use std::time::{SystemTime, UNIX_EPOCH};
use bytes::{Bytes, BytesMut, BufMut, Buf};
use crc32fast;

/// Protocol magic number (4 bytes) - "UTPB" (Universal Transport Protocol Binary)
pub const PROTOCOL_MAGIC: u32 = 0x55545042;

/// Protocol version (1 byte)
pub const PROTOCOL_VERSION: u8 = 1;

/// Message header size (32 bytes - cache line aligned)
pub const HEADER_SIZE: usize = 32;

/// Maximum payload size (64MB)
pub const MAX_PAYLOAD_SIZE: u32 = 64 * 1024 * 1024;

/// Message types (1 byte)
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MessageType {
    Data = 0x01,
    Heartbeat = 0x02,
    Acknowledgment = 0x03,
    Error = 0x04,
    Benchmark = 0x05,
}

impl From<u8> for MessageType {
    fn from(value: u8) -> Self {
        match value {
            0x01 => MessageType::Data,
            0x02 => MessageType::Heartbeat,
            0x03 => MessageType::Acknowledgment,
            0x04 => MessageType::Error,
            0x05 => MessageType::Benchmark,
            _ => MessageType::Data, // Default fallback
        }
    }
}

/// Binary message header (32 bytes, fixed layout)
/// Layout:
/// 0-3:   Magic number (4 bytes)
/// 4:     Version (1 byte)
/// 5:     Message type (1 byte)
/// 6-7:   Flags (2 bytes)
/// 8-11:  Payload length (4 bytes)
/// 12-19: Sequence number (8 bytes)
/// 20-27: Timestamp (8 bytes, microseconds since epoch)
/// 28-31: CRC32 checksum (4 bytes)
#[repr(C, packed)]
#[derive(Debug, Clone, Copy)]
pub struct BinaryHeader {
    pub magic: u32,
    pub version: u8,
    pub message_type: u8,
    pub flags: u16,
    pub payload_length: u32,
    pub sequence: u64,
    pub timestamp: u64,
    pub checksum: u32,
}

impl BinaryHeader {
    /// Create a new binary header
    pub fn new(message_type: MessageType, payload: &[u8]) -> Self {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_micros() as u64;
        
        let checksum = Self::calculate_crc32(payload);
        
        Self {
            magic: PROTOCOL_MAGIC,
            version: PROTOCOL_VERSION,
            message_type: message_type as u8,
            flags: 0,
            payload_length: payload.len() as u32,
            sequence: 0, // Set by sender
            timestamp,
            checksum,
        }
    }
    
    /// Set sequence number
    pub fn set_sequence(&mut self, seq: u64) {
        self.sequence = seq;
    }
    
    /// Validate header
    pub fn validate(&self) -> Result<(), ProtocolError> {
        if self.magic != PROTOCOL_MAGIC {
            return Err(ProtocolError::InvalidMagic(self.magic));
        }
        
        if self.version != PROTOCOL_VERSION {
            return Err(ProtocolError::UnsupportedVersion(self.version));
        }
        
        if self.payload_length > MAX_PAYLOAD_SIZE {
            return Err(ProtocolError::PayloadTooLarge(self.payload_length));
        }
        
        Ok(())
    }
    
    /// Verify payload checksum
    pub fn verify_checksum(&self, payload: &[u8]) -> bool {
        self.checksum == Self::calculate_crc32(payload)
    }
    
    /// Calculate CRC32 checksum
    fn calculate_crc32(data: &[u8]) -> u32 {
        crc32fast::hash(data)
    }
    
    /// Serialize header to bytes (little-endian)
    pub fn to_bytes(&self) -> [u8; HEADER_SIZE] {
        let mut bytes = [0u8; HEADER_SIZE];
        let mut buf = BytesMut::with_capacity(HEADER_SIZE);
        
        buf.put_u32_le(self.magic);
        buf.put_u8(self.version);
        buf.put_u8(self.message_type);
        buf.put_u16_le(self.flags);
        buf.put_u32_le(self.payload_length);
        buf.put_u64_le(self.sequence);
        buf.put_u64_le(self.timestamp);
        buf.put_u32_le(self.checksum);
        
        bytes.copy_from_slice(&buf[..HEADER_SIZE]);
        bytes
    }
    
    /// Deserialize header from bytes
    pub fn from_bytes(bytes: &[u8]) -> Result<Self, ProtocolError> {
        if bytes.len() < HEADER_SIZE {
            return Err(ProtocolError::InsufficientData(bytes.len()));
        }
        
        let mut buf = Bytes::copy_from_slice(&bytes[..HEADER_SIZE]);
        
        let header = Self {
            magic: buf.get_u32_le(),
            version: buf.get_u8(),
            message_type: buf.get_u8(),
            flags: buf.get_u16_le(),
            payload_length: buf.get_u32_le(),
            sequence: buf.get_u64_le(),
            timestamp: buf.get_u64_le(),
            checksum: buf.get_u32_le(),
        };
        
        header.validate()?;
        Ok(header)
    }
}

/// Complete binary message
#[derive(Debug, Clone)]
pub struct BinaryMessage {
    pub header: BinaryHeader,
    pub payload: Bytes,
}

impl BinaryMessage {
    /// Create a new binary message
    pub fn new(message_type: MessageType, payload: Bytes) -> Result<Self, ProtocolError> {
        if payload.len() > MAX_PAYLOAD_SIZE as usize {
            return Err(ProtocolError::PayloadTooLarge(payload.len() as u32));
        }
        
        let header = BinaryHeader::new(message_type, &payload);
        
        Ok(Self { header, payload })
    }
    
    /// Create a benchmark message with specific data
    pub fn benchmark(id: u64, data: Bytes) -> Result<Self, ProtocolError> {
        let mut message = Self::new(MessageType::Benchmark, data)?;
        message.header.set_sequence(id);
        Ok(message)
    }
    
    /// Get total message size
    pub fn total_size(&self) -> usize {
        HEADER_SIZE + self.payload.len()
    }
    
    /// Validate complete message
    pub fn validate(&self) -> Result<(), ProtocolError> {
        self.header.validate()?;
        
        if !self.header.verify_checksum(&self.payload) {
            return Err(ProtocolError::ChecksumMismatch);
        }
        
        Ok(())
    }
    
    /// Serialize entire message to bytes
    pub fn to_bytes(&self) -> Bytes {
        let total_size = self.total_size();
        let mut buf = BytesMut::with_capacity(total_size);
        
        // Header
        buf.put_slice(&self.header.to_bytes());
        
        // Payload
        buf.put_slice(&self.payload);
        
        buf.freeze()
    }
    
    /// Deserialize message from bytes
    pub fn from_bytes(bytes: &[u8]) -> Result<Self, ProtocolError> {
        if bytes.len() < HEADER_SIZE {
            return Err(ProtocolError::InsufficientData(bytes.len()));
        }
        
        // Parse header
        let header = BinaryHeader::from_bytes(bytes)?;
        
        // Check if we have enough data for payload
        let expected_total = HEADER_SIZE + header.payload_length as usize;
        if bytes.len() < expected_total {
            return Err(ProtocolError::InsufficientData(bytes.len()));
        }
        
        // Extract payload
        let payload = Bytes::copy_from_slice(&bytes[HEADER_SIZE..expected_total]);
        
        let message = Self { header, payload };
        message.validate()?;
        
        Ok(message)
    }
}

/// Benchmark-specific message structure
#[derive(Debug, Clone)]
pub struct BenchmarkMessage {
    pub id: u64,
    pub timestamp: u64,
    pub data: Bytes,
    pub metadata: String,
}

impl BenchmarkMessage {
    /// Create a new benchmark message
    pub fn new(id: u64, data_size: usize) -> Self {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_micros() as u64;
        
        Self {
            id,
            timestamp,
            data: Bytes::from(vec![0x42; data_size]),
            metadata: format!("benchmark_msg_{}", id),
        }
    }
    
    /// Serialize to binary format (fixed layout)
    /// Layout:
    /// 0-7:    ID (8 bytes)
    /// 8-15:   Timestamp (8 bytes)
    /// 16-19:  Data length (4 bytes)
    /// 20-23:  Metadata length (4 bytes)
    /// 24+:    Data
    /// N+:     Metadata (UTF-8)
    pub fn to_binary(&self) -> Bytes {
        let metadata_bytes = self.metadata.as_bytes();
        let total_size = 24 + self.data.len() + metadata_bytes.len();
        
        let mut buf = BytesMut::with_capacity(total_size);
        
        buf.put_u64_le(self.id);
        buf.put_u64_le(self.timestamp);
        buf.put_u32_le(self.data.len() as u32);
        buf.put_u32_le(metadata_bytes.len() as u32);
        buf.put_slice(&self.data);
        buf.put_slice(metadata_bytes);
        
        buf.freeze()
    }
    
    /// Deserialize from binary format
    pub fn from_binary(bytes: &[u8]) -> Result<Self, ProtocolError> {
        if bytes.len() < 24 {
            return Err(ProtocolError::InsufficientData(bytes.len()));
        }
        
        let mut buf = Bytes::copy_from_slice(bytes);
        
        let id = buf.get_u64_le();
        let timestamp = buf.get_u64_le();
        let data_len = buf.get_u32_le() as usize;
        let metadata_len = buf.get_u32_le() as usize;
        
        // Check if we have enough remaining data
        if buf.remaining() < data_len + metadata_len {
            return Err(ProtocolError::InsufficientData(buf.remaining()));
        }
        
        let data = buf.split_to(data_len);
        let metadata_bytes = buf.split_to(metadata_len);
        let metadata = String::from_utf8(metadata_bytes.to_vec())
            .map_err(|_| ProtocolError::InvalidUtf8)?;
        
        Ok(Self {
            id,
            timestamp,
            data,
            metadata,
        })
    }
    
    /// Convert to binary message
    pub fn to_binary_message(&self) -> Result<BinaryMessage, ProtocolError> {
        let payload = self.to_binary();
        BinaryMessage::benchmark(self.id, payload)
    }
    
    /// Create from binary message
    pub fn from_binary_message(msg: &BinaryMessage) -> Result<Self, ProtocolError> {
        Self::from_binary(&msg.payload)
    }
}

/// Protocol errors
#[derive(Debug, thiserror::Error)]
pub enum ProtocolError {
    #[error("Invalid magic number: 0x{0:x}")]
    InvalidMagic(u32),
    
    #[error("Unsupported protocol version: {0}")]
    UnsupportedVersion(u8),
    
    #[error("Payload too large: {0} bytes")]
    PayloadTooLarge(u32),
    
    #[error("Insufficient data: {0} bytes available")]
    InsufficientData(usize),
    
    #[error("Checksum mismatch")]
    ChecksumMismatch,
    
    #[error("Invalid UTF-8 encoding")]
    InvalidUtf8,
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_header_serialization() {
        let payload = b"Hello, World!";
        let header = BinaryHeader::new(MessageType::Data, payload);
        
        let bytes = header.to_bytes();
        let decoded = BinaryHeader::from_bytes(&bytes).unwrap();
        
        assert_eq!(header.magic, decoded.magic);
        assert_eq!(header.message_type, decoded.message_type);
        assert_eq!(header.payload_length, decoded.payload_length);
        assert!(decoded.verify_checksum(payload));
    }
    
    #[test]
    fn test_benchmark_message_serialization() {
        let msg = BenchmarkMessage::new(123, 1024);
        let binary = msg.to_binary();
        let decoded = BenchmarkMessage::from_binary(&binary).unwrap();
        
        assert_eq!(msg.id, decoded.id);
        assert_eq!(msg.data.len(), decoded.data.len());
        assert_eq!(msg.metadata, decoded.metadata);
    }
    
    #[test]
    fn test_complete_message_flow() {
        let bench_msg = BenchmarkMessage::new(456, 512);
        let binary_msg = bench_msg.to_binary_message().unwrap();
        let serialized = binary_msg.to_bytes();
        
        let deserialized = BinaryMessage::from_bytes(&serialized).unwrap();
        let recovered_bench = BenchmarkMessage::from_binary_message(&deserialized).unwrap();
        
        assert_eq!(bench_msg.id, recovered_bench.id);
        assert_eq!(bench_msg.data.len(), recovered_bench.data.len());
    }
}